src/fetching/provider/any.rs

```diff
@@ -0,0 +1,254 @@
+///
+/// This provider adaptor implements the [`Provider`](super::Provider) protocol by fetching
+/// requested objects from several different underlying providers. If any of the underlying sources
+/// have the object, the request will eventually succeed.
+///
+/// This can be used to combine multiple instances of the same kind of provider, like using
+/// [`QueryServiceProvider`](super::QueryServiceProvider) to request objects from a number of
+/// different query services. It can also be used to search different kinds of data providers for
+/// the same object, like searching for a block both in another instance of the query service and in
+/// the HotShot DA committee. Finally, [`AnyProvider`] can be used to combine a provider which only
+/// provides blocks and one which only provides leaves into a provider which provides both, and thus
+/// can be used as a provider for the availability API module.
+///
+/// # Examples
+///
+/// Fetching from multiple query services, for resiliency.
+///
+/// ```
+/// # use hotshot_types::traits::node_implementation::NodeType;
+/// # async fn doc<Types>() -> anyhow::Result<()>
+/// # where
+/// #   Types: NodeType,
+/// # {
+/// use hotshot_query_service::fetching::provider::{AnyProvider, QueryServiceProvider};
+///
+/// let qs1 = QueryServiceProvider::new("https://backup.query-service.1".parse()?);
+/// let qs2 = QueryServiceProvider::new("https://backup.query-service.2".parse()?);
+/// let provider = AnyProvider::<Types>::default()
+///     .with_provider(qs1)
+///     .with_provider(qs2);
+/// # Ok(())
+/// # }
+/// ```
+#[derive(Derivative)]
+#[derivative(Clone(bound = ""), Debug(bound = ""), Default(bound = ""))]
+pub struct AnyProvider<Types>
+where
+    Types: NodeType,
+{
+    payload_providers: Vec<PayloadProvider<Types>>,
+    leaf_providers: Vec<LeafProvider<Types>>,
+}
+
+#[async_trait]
+impl<Types> Provider<Types, PayloadRequest> for AnyProvider<Types>
+where
+    Types: NodeType,
+{
+    async fn fetch(&self, req: PayloadRequest) -> Option<Payload<Types>> {
+        any_fetch(&self.payload_providers, req).await
+    }
+}
+
+#[async_trait]
+impl<Types> Provider<Types, LeafRequest> for AnyProvider<Types>
+where
+    Types: NodeType,
+{
+    async fn fetch(&self, req: LeafRequest) -> Option<LeafQueryData<Types>> {
+        any_fetch(&self.leaf_providers, req).await
+    }
+}
+
+impl<Types> AnyProvider<Types>
+where
+    Types: NodeType,
+{
+    /// Add a sub-provider which fetches both blocks and leaves.
+    pub fn with_provider<P>(mut self, provider: P) -> Self
+    where
+        P: Provider<Types, PayloadRequest> + Provider<Types, LeafRequest> + Debug + 'static,
+    {
+        let provider = Arc::new(provider);
+        self.payload_providers.push(provider.clone());
+        self.leaf_providers.push(provider);
+        self
+    }
+
+    /// Add a sub-provider which fetches blocks.
+    pub fn with_block_provider<P>(mut self, provider: P) -> Self
```

Comment:
I'm curious, how would a generic version of this look, I am spacey right now with Rust generics (thanks Go lol), but would it be possible to have some function like `with_provider<P, T>` that adds to a generic providers map like:
```
providers: Vec<HashMap<TypeKey<T>, P> 
```

Anyways, this isn't important because we only have two sub-providers but I was curious.

Reply:
Yeah, I think you could do something like this with [`TypeId`]. But it would be messy, since you don't know the types of `T` and `P` at the data structure level (only in this particular method) you would have to do something like `HashMap<TypeId, Arc<dyn Provider>>`, and then when you want to use the provider you would have to downcast it to a specific request type.